org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -312,30 +306,46 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
 		}
 
 		final ToBeBuilt toBeBuilt = new ToBeBuilt();
-		IResourceDeltaVisitor visitor = new IResourceDeltaVisitor() {
-			@Override
-			public boolean visit(IResourceDelta delta) throws CoreException {
-				if (progress.isCanceled())
-					throw new OperationCanceledException();
-				if (delta.getResource() instanceof IProject) {
-					return delta.getResource() == getProject();
-				}
-				if (delta.getResource() instanceof IStorage) {
-					if (delta.getKind() == IResourceDelta.REMOVED) {
-						return toBeBuiltComputer.removeStorage(null, toBeBuilt, (IStorage) delta.getResource());
-					} else if (delta.getKind() == IResourceDelta.ADDED || delta.getKind() == IResourceDelta.CHANGED) {
-						return toBeBuiltComputer.updateStorage(null, toBeBuilt, (IStorage) delta.getResource());
-					}
-				}
-				return true;
-			}
-		};
+		IResourceDeltaVisitor visitor = createDeltaVisitor(toBeBuiltComputer, toBeBuilt, progress);
 		delta.accept(visitor);
+		processedAbsentReferencedProjects(visitor);
 		if (progress.isCanceled())
 			throw new OperationCanceledException();
 		progress.worked(2);
 		doBuild(toBeBuilt, progress.split(8), BuildType.INCREMENTAL);
 	}
+	
+	/**
+	 * Obtain the deltas for the projects we were previously interested in and process them too, if they do no longer
+	 * have the xtext nature or if they are no longer accessible.
+	 *
+	 * @param visitor
+	 *            the processor.
+	 * @throws CoreException
+	 *             if something goes down the drain.
+	 * @since 2.17
+	 */
+	protected void processedAbsentReferencedProjects(IResourceDeltaVisitor visitor) throws CoreException {
+		final IProject[] interestingProjects = optimisticInvoke("getInterestingProjects");
```

Comment:
For the stuff that we invoke reflectively here, we should file a change request against platform.